docs: bootloader: Add a user guide (#3607)
nrfconnect/sdk-nrf: doc/nrf/ug_bootloader.rst

  This is useful when a bootloader performs additional complex operations, like using a software stack.

  See below for an abstract representation of the memory layout for an application that uses both an immutable and an upgradable bootloader in its boot chain:

  .. figure:: images/bootloader_memory_layout.svg
     :alt: Memory layout

For detailed information about the memory layout, see the partition configuration in the :file:`partitions.yml` file, located in the build folder directory, or run ``ninja partition_manager_report``.

.. note::
   You must enable the Partition Manager to make the :file:`partitions.yml` file and the ``partition_manager_report`` target available.
   See :ref:`partition_manager` and :ref:`ug_multi_image` for details.

By default, building an application with any bootloader configuration creates a :ref:`multi-image build <ug_multi_image>`, where the :ref:`partition_manager` manages its memory partitions.
When building with a non-secure application, the :ref:`secure_partition_manager` is also built with the image automatically.

.. _immutable_bootloader:

> First-stage Immutable bootloader

Author: Fixed.

   :local:
   :depth: 2

The core architecture behind the secure bootloader solution provided by |NCS| is based on the *chain of trust* concept.

> By using this secure bootloader chain, you can easily ensure that all code that is executed is authorized and your application is protected against running altered code.

Reviewer: ```suggestion
By using this secure bootloader chain, you can ensure on the one hand that all code that is executed is authorized and, on the other hand, that your application is protected against running altered code.
```
Had to reread this three times because of the concatenation of "that" and "is", so I'm suggesting a small change.

Author: Reworded from your suggestion. Please see latest commit.

If, for example, an attacker attempts to modify your application, or you have a bug in your code that overwrites parts of the application firmware image, the secure bootloader chain detects that the firmware has been altered and your application does not start up.

> You should choose the structure of the secure bootloader chain depending on your desired firmware update scheme:

Reviewer: ```suggestion
When choosing the structure of the secure bootloader chain, make the decision based on your desired firmware update scheme:
```
`should` sounds like `you can, but you don't have to`, and I don't know if this is what is intended here.

Reviewer 2: I mean whether need to update application  or app and mcuboot.

Author: Ah yes. This is mentioned in the bullets following this statement.

Author: Ah yes, this is brought up in the bullets following this statement.

Each step in this chain guarantees that the next step can be trusted to have certain properties because any unauthorized modification of a subsequent step will be detected and the process halted.
The trustworthiness of each layer is guaranteed by the previous layer, all the way back to a property in the system referred to as *root of trust* (RoT).

You can compare a chain of trust to the concept of a door:

1. You trust a door because you trust the lock.
#. You trust the lock because you trust the key.
#. You trust the key because it is in your pocket.

If you lose this key, the chain of trust unravels and you can no longer trust this door.

It is important to note that a key alone is not a RoT, although a RoT may include a key.
A RoT will consist of hardware, software, and data components that must always behave as expected because any misbehavior cannot be detected.

.. _ug_bootloader_architecture:

Architecture
************

In the context of the bootloader, a step in the chain of trust is the boot of a specific image.
As such, the secure bootloader chain consists of a sequence of images that are booted one after the other.

> To establish a root of trust, the first image verifies the signature of the next image, which can either be an application or another bootloader image.

Reviewer: What is `a root of trust` and how is it different than `a chain of trust`?

Author: This is covered a bit a little earlier in the Chain of Trust section:

> The trustworthiness of each layer is guaranteed by the previous layer, all the way back to a property in the system referred to as *root of trust* (RoT)...It is important to note that a key alone is not a RoT, although a RoT may include a key.
A RoT will consist of hardware, software, and data components that must always behave as expected because any misbehavior cannot be detected.

But there is no succinct definition anywhere. Since this is an external term used with things like [PSA](https://www.psacertified.org/blog/what-is-a-root-of-trust/), I'm not sure its within our scope to define it, but we could perhaps link to PSA's definition for reference?

If the next image is another bootloader image, that one must verify the image following it to maintain the chain of trust.

> After all of the images in the bootloader chain have been verified successfully, the application starts.

Reviewer: Wouldn't this information be more suitable in the previous section about Chain of trust? I feel architecture discussion starts only in line 52.

Author: Good point. I moved it up into the Chain of Trust section and reworded it a bit to aid with the flow. Please see the latest commit.

.. _upgradable_bootloader:

Second-stage Upgradable bootloader
==================================

If you also need the capability of updating the bootloader, you can use a second-stage, upgradable bootloader.
It can be updated through :ref:`firmware-over-the-air (FOTA) <lib_fota_download>` updates, unlike the immutable bootloader.

The root of trust, in the form of the immutable bootloader, protects the upgradable bootloader, which must also continue the chain of trust by verifying the next image in the boot sequence.
Because of this, the immutable bootloader is responsible for upgrading the upgradable bootloader and verifying its metadata and image integrity.
For more information about how the immutable bootloader accomplishes this, see the :ref:`bootloader_flash_layout` section of the |NSIB|.

The upgradable bootloader carries out the same signature and metadata verification as the immutable bootloader.
Also, it can upgrade both itself and the following image in the boot sequence, which, in most cases, is an application.

.. caution::
   You should add a second-stage bootloader only when necessary by the design or firmware upgrade needs.

>    Adding the second stage bootloader for no reason will lead to a degradation of the system's overall security, as attackers can exploit bugs that may exist in either bootloader.

Reviewer 2: very probably overall security level estimation of `NISB + sdk-mcuboot` is higher than standalone sdk-mcuboot as 2-stages solution gives capability of upgrade sdk-mcuboot (which has much bigger codebase than NISB and therefore might be more affected by bugs).

Author: I see your point here, but I think what @pdunaj was meaning by this caution is that more code in general means higher chance of bugs, even if one of those codebases is upgradable.

* Signature verification - Verifies that the key used for signing the next image in the boot sequence matches one of the provided public keys.
* Metadata verification - Checks that the images are compatible.

.. note::

   It is *critical* to generate and use your own signing keys while in development and before deploying when using either MCUboot or the |NSIB| as an immutable bootloader.
   See :ref:`ug_fw_update_development_keys` for more information.

The :ref:`bootloader capabilities table <app_bootloaders_support_table>` indicates the bootloaders that are valid for use as immutable bootloaders.

.. _upgradable_bootloader:

Second-stage Upgradable bootloader
==================================

A second-stage, upgradable bootloader can be used if you also need the capability of updating the bootloader.
It can be updated through :ref:`firmware-over-the-air (FOTA) <lib_fota_download>` updates, unlike the immutable bootloader.

The upgradable bootloader is protected by the root of trust in the form of the immutable bootloader, and it must continue the chain of trust by verifying the next image in the boot sequence.

> Because of this, the immutable bootloader is responsible for upgrading the

Author: Fixed in latest commit.

Having both slots programmed allows the immutable bootloader to invalidate the corrupt image and boot into a valid one.

.. _ug_bootloader_flash:

Flash memory partitions
=======================

Flash memory partitioning is handled differently for each bootloader.

A report of how the flash partitioning has been handled for a bootloader or combination of bootloaders can be printed image with :ref:`pm_partition_reports` after building the application.

.. _ug_bootloader_flash_b0:

|NSIB| partitions
-----------------

See :ref:`bootloader_flash_layout` for implementation-specific information about this bootloader.

.. _ug_bootloader_flash_mcuboot:

MCUboot partitions
------------------

For most applications, MCUboot requires two image slots:

* The *primary slot*, containing the application that will be booted.
* The *secondary slot*, where a new application can be stored before it is activated.

It is possible to use only a single *primary slot* for MCUboot by using the ``CONFIG_SINGLE_APPLICATION_SLOT`` option.

> This is particularly useful in memory-constrained devices to limit the flash allocated for the bootloader.

Author: Updated in latest commit.

.. note::
   You must enable the Partition Manager to make the :file:`partitions.yml` file and the ``partition_manager_report`` target available.
   See :ref:`partition_manager` and :ref:`ug_multi_image` for details.

By default, building an application with any bootloader configuration creates a :ref:`multi-image build <ug_multi_image>`, where the :ref:`partition_manager` manages its memory partitions.
When building with a non-secure application, the :ref:`secure_partition_manager` is also built with the image automatically.

.. _immutable_bootloader:

First-stage Immutable bootloader
================================

The first step in the chain of trust is a secure, immutable, first-stage bootloader.

This immutable bootloader runs after every reset and establishes the root of trust by verifying the signature and metadata of the next image in the boot sequence.
If the verification fails, the boot process stops.
This way, the immutable bootloader can guarantee that the next image in the boot sequence will not start up if it has been tampered with in any way.

> For example, if an attacker attempts to take over the device by altering the firmware, the device will not boot, and thus not run the infected code.

Reviewer 3: There is one important thing to note here. I guess it is obvious but for this to work bootloader must be truly immutable.
In the next paragraph it states

> immutable bootloader cannot be updated or deleted from the device unless you erase the device

The whole point however is that bootloader should not be modifiable in any way and that means that erase must be locked out (in HW if possible).

If one can do something to erase and change the bootloader (e.g. by connecting the debug pins), it means the whole chain of trust does not exist.

If process of locking out the flash is mentioned in some place then link to it should be provided. If it is not I believe it should be described.

Author: The locked flash is mentioned in the NSIB README, which I've now linked along with some wording changes.

The immutable bootloader cannot be updated or deleted from the device unless you erase the device.

> There is no need to modify the immutable bootloader in any way before you program it.

Reviewer 4: ```suggestion
Except for providing your own public keys, there is no need to modify the immutable bootloader in any way before you program it.
```

Author: Updated in latest commit.

.. _upgradable_bootloader_presigned_variants:

Pre-signed variants
-------------------

When programming an upgradable bootloader, the build system is able to automatically generate pre-signed variants of the image for both slots when compiling with the :option:`CONFIG_BUILD_S1_VARIANT` option.
An OTA update package of the upgradable bootloader contains both images, since it might be unknown which slot is available for upgrade.
See the :ref:`bootloader_flash_layout` of the |NSIB| for more details.

Therefore, these variants are necessary to perform the upgrade procedure of the bootloader through the :ref:`lib_fota_download` library.

Additionally, building and programming these variants acts as a backup mechanism in case the image in the primary slot becomes corrupted, for example from a bit-flip, as the immutable bootloader can invalidate the corrupt image and boot into a valid one.

.. note::

   Building a pre-signed variant for slot 1 will not change the flash memory partition layout for an application, as the slot is already taken into account by default.

>    Instead, the flash memory content of slot 1 in the built hex file will be populated by an image instead of being erased or left as-is.

Reviewer 4: I don't 100% understand the meaning of this. Who doesn't change the partition layout, and who populates the slot?

Author: Agreed, the wording needed to be updated. I changed up this whole section about pre-signed variants and moved some relevant information into the NSIB README. Please have a look at the latest version and see if it is better about this.

   :local:
   :depth: 2

The |NCS| provides a secure bootloader solution based on the *chain of trust* concept.

By using this bootloader chain, you can easily ensure that all code that is executed is authorized and your application is protected against running altered code.
If, for example, an attacker attempts to modify your application, or you have a bug in your code that overwrites parts of the application firmware image, the secure bootloader chain detects that the firmware has been altered and your application does not start up.

Chain of trust
**************

A secure system depends on building and maintaining a chain of trust through all layers in the system.

> Each step in this chain will guarantee that the next step can be trusted to have certain properties, because any illegal modification of a subsequent step will be detected and the process halted.

Reviewer 5: Added a clarifications at the start of the next chapter

1. You trust a door because you trust the lock.
#. You trust the lock because you trust the key.
#. You trust the key because it is in your pocket.

If you lose this key the chain of trust unravels and you can no longer trust this door.

It is important to note that a key alone is not a RoT, although a RoT may include a key.
A RoT will consist of hardware, software, and data components that must always behave as expected because any misbehavior cannot be detected.

.. _ug_bootloader_architecture:

Architecture
************

The secure bootloader chain consists of a sequence of images that are booted one after the other.
To establish a root of trust, the first image verifies the signature of the next image, which can either be the application or another bootloader image.
If the next image is another bootloader image, that one must verify the image following it to maintain the chain of trust.
After all of the images in the bootloader chain have been verified successfully, the application starts.

The current implementation provides the first stage in the chain, the :ref:`bootloader`, and uses :doc:`mcuboot:index` as upgradable bootloader.

> The following image shows an abstract representation of the memory layout for an application that uses an immutable and upgradable bootloader in its boot chain:

Reviewer 5: @hakonfam Updated, please check also the new picture.

When building with a |trusted_execution| application, the :ref:`secure_partition_manager` is also built with the image automatically.

.. _immutable_bootloader:

Immutable bootloader
====================

The first step in the chain of trust is a secure, immutable, first-stage bootloader, referred to as the *immutable bootloader*.

The immutable bootloader runs after every reset and establishes the root of trust by verifying the signature and metadata of the next image in the boot sequence.
If the verification fails, the boot process stops.
This way, the immutable bootloader can guarantee that the next image in the boot sequence will not start up if it has been tampered with in any way.
For example, if an attacker attempts to take over the device by altering the firmware, the device will not boot, and thus not run the infected code.

The immutable bootloader cannot be updated or deleted from the device unless you erase the device.

There is no need to modify the immutable bootloader in any way before you program it.

> The default verification is recommended and suitable for all the most common user scenarios and includes the following checks:

Reviewer 4: Yes, this is very very important. For both MCUboot and B0, they MUST generate their own keys before deploying.

Of course they can try it without modifying :)

> The default verification is recommended and suitable for all the most common user scenarios and includes the following checks:

Author: Agreed. Both `ug_fw_update.rst` and `ug_bootloader.rst` now contain information highlighting the importance of this, with the latter referencing the former for more information (it is presented as a warning to be more attention grabbing).

   Verifies that the key used for signing the next image in the boot sequence matches one of the provided public keys.

Metadata verification
   Checks that the images are compatible.

The :ref:`bootloader capabilities table <app_bootloaders_support_table>` details which bootloaders are valid to use as an immutable bootloader.

.. _upgradable_bootloader:

Upgradable bootloader
=====================

Unlike the immutable bootloader, the *upgradable bootloader* can be updated through :ref:`firmware-over-the-air (FOTA) <lib_fota_download>` updates.
Therefore, this bootloader is more flexible than the immutable bootloader.
It is protected by the root of trust in form of the immutable bootloader, and it must continue the chain of trust by verifying the next image in the boot sequence.

The upgradable bootloader should carry out the same signature and metadata verification as the immutable bootloader.
In addition, it can provide functionality for upgrading both itself and the following image in the boot sequence, which, in most cases, is the application.

The immutable bootloader supports two partition slots in which the upgradable bootloader can be stored: slot 0 and slot 1.

> Immutable bootloaders always update the upgradable bootloader using the :doc:`direct-xip <mcuboot:design>` strategy: placing a new bootloader image in the slot that is not in use by the current version of the upgradable bootloader.

Author: Reworded this in the latest commit, and moved some wording around regarding this and the NSIB documentation. It should be more clear about what this is doing and _not_ leveraging MCUboot's direct-xip.

When booting, the immutable bootloader checks the version information for the images in slot 0 and slot 1 and boots the one with the highest version.
If this image is faulty and cannot be booted, the other partition will always hold a working image that will be booted instead.

The :ref:`bootloader capabilities table <app_bootloaders_support_table>` details which bootloaders are valid to use as an upgradable bootloader.

.. _upgradable_bootloader_presigned_variants:

Pre-signed variants
-------------------

> When programming an upgradable bootloader, it is possible to automatically generate pre-signed variants of the image for both slots.

Author: Added. Please see latest commit.

> This is to ensure that the update candidate is linked correctly against both image slots, as it is not guaranteed to be executed from one particular slot.

Reviewer 4: ```suggestion
An OTA update package of the upgradable bootloader contains both images, since it might be unknown which slot is available for upgrade.
```

Therefore, these variants are necessary to perform the upgrade procedure of the bootloader through the :ref:`lib_fota_download` library.

Additionally, building and programming these variants acts as a backup mechanism in case the image in the primary slot becomes corrupted, for example from a bit-flip, in which case the immutable bootloader can invalidate the corrupt image and boot into a valid one.

.. note::

>    Including a pre-signed variant in slot 1 will not change the flash partition layout for an application, as the slot is already taken into account by default.

Author: My original intention was the hex file. I have updated the wording to clarify this.

* The *secondary slot*, where a new application can be stored before it is activated.

See the *Image Slots* section in the :doc:`MCUboot documentation <mcuboot:design>` for more information.

The |NCS| variant of MCUboot uses the :ref:`partition_manager` to configure the flash partitions for these image slots.
In the default configuration, defined in :file:`bootloader/mcuboot/boot/zephyr/pm.yml`, the partition manager dynamically sets up the partitions as required for MCUboot.
For example, the partition layout for :file:`zephyr/samples/hello_world` using MCUboot on the ``nrf52840dk_nrf52840`` development kit would look like the following:

.. NOTE: The following ASCII blocks render strangely if the first character before the parenthesis is not a whitespace

.. code-block:: console

    (0x100000 - 1024.0kB):
   +-----------------------------------------+
   | 0x0: mcuboot (0xc000)                   |
   +---0xc000: mcuboot_primary (0x7a000)-----+
   | 0xc000: mcuboot_pad (0x200)             |
   +---0xc200: mcuboot_primary_app (0x79e00)-+
   | 0xc200: app (0x79e00)                   |
   | 0x86000: mcuboot_secondary (0x7a000)    |
   +-----------------------------------------+

> If you want to control where in memory the flash partitions are placed, you can define :ref:`static partitions <ug_pm_static>` for your application.

Reviewer 6: You should change the kconfig to get the sizes as you want them. 

You only use static partitions after you have shipped a product so that you can no longer change the partition layout.

Author: When [relocating this section to partition manager docs](https://github.com/nrfconnect/sdk-nrf/pull/3607#discussion_r607163773), I removed this sentence. I wasn't sure how best to word it with extensive docs about this already being present.

Additionally, it is possible to store secondary slot images in an external flash memory when using MCUboot.
See :ref:`ug_bootloader_adding_external_flash` for more information.

Partition Reports
-----------------

> To get a better overview of how flash partitions are created by the Partition Manager, especially when using multiple bootloaders, it is possible to generate ASCII-art map reports using the :file:`nrf/scripts/partition_manager_report.py` script:

Reviewer 6: You should invoke the `ninja partition_manager_report` instead. Change that below as well.

Author: Changed. Please see latest commit.

.. code-block:: console

   python3 nrf/scripts/partition_manager_report.py -i build/partitions.yml

For example, calling this on the build of ``zephyr/samples/hello_world`` with the nRF52840 development kit using |b0| and MCUboot, the ASCII partition output would look like the following:

.. code-block:: console

    (0x100000 - 1024.0kB):
   +------------------------------------------+

>    +---0x0: b0 (0x9000)-----------------------+

Reviewer 6: Should we state somewhere where these are defined? I mean point to the .yml files?

Author: Added some text to clarify this. Please see latest commit.

.. code-block:: console

    (0x100000 - 1024.0kB):
   +-----------------------------------------+
   | 0x0: mcuboot (0xc000)                   |
   +---0xc000: mcuboot_primary (0x7a000)-----+
   | 0xc000: mcuboot_pad (0x200)             |
   +---0xc200: mcuboot_primary_app (0x79e00)-+
   | 0xc200: app (0x79e00)                   |
   | 0x86000: mcuboot_secondary (0x7a000)    |
   +-----------------------------------------+

If you want to control where in memory the flash memory partitions are placed, you can define :ref:`static partitions <ug_pm_static>` for your application.

Additionally, it is possible to store secondary slot images in external flash memory when using MCUboot.
See :ref:`ug_bootloader_adding_external_flash` for more information.

Partition Reports
-----------------

To get a better overview of how flash memory partitions are created by the Partition Manager, especially when using multiple bootloaders, it is possible to generate ASCII-art map reports using the :file:`nrf/scripts/partition_manager_report.py` script:

.. code-block:: console

   python3 nrf/scripts/partition_manager_report.py -i build/partitions.yml

> For example, calling this on the build of ``zephyr/samples/hello_world`` with the nRF52840 development kit using |b0| and MCUboot, the ASCII partition output would look like the following:

Reviewer 4: You could mention that these reports are printed when calling `ninja rom_report`.

Author: This whole section seemed much more relevant to partition manager docs anyway, so I moved it to `partition_manager.rst` as its own section about generating reports. Please see latest commit.

To establish a root of trust, the first image verifies the signature of the next image, which can either be an application or another bootloader image.
If the next image is another bootloader image, that one must verify the image following it to maintain the chain of trust.
After all of the images in the bootloader chain have been verified successfully, the application starts.

There are two implementations currently supported:

* The first implementation provides the first stage in the chain, the immutable :ref:`bootloader`, which could be either |NSIB| or :doc:`mcuboot:index`.
  It does not support bootloader upgradability.
  It is useful if you need just the capability to update your application.

  The following image shows an abstract representation of the memory layout for an application that uses only an immutable bootloader in its boot chain:

  .. figure:: images/bootloader_memory_layout_onestage.svg
     :alt: Memory layout

* The second implementation provides both the first stage in the chain, the immutable bootloader (|NSIB|), and uses :doc:`mcuboot:index` as the upgradable second-stage bootloader.

>   This implementation is useful if you need the capability to update both your bootloader and your application.

Reviewer 3: > update both your bootloader

Maybe adding a hit why to do it would be helpful? e.g. "when bootloader performs additional complex operations, use SW stacks, etc". Adding the second stage bootloader for no reason will lead to degradation of the overall security as attacker can exploit bugs in either of bootloaders.

Reviewer 5: @pdunaj rephrased, please check the updated version

Therefore, these variants are necessary to perform the upgrade procedure of the bootloader through the :ref:`lib_fota_download` library.

Additionally, building and programming these variants acts as a backup mechanism in case the image in the primary slot becomes corrupted, for example from a bit-flip, as the immutable bootloader can invalidate the corrupt image and boot into a valid one.

.. note::

   Including a pre-signed variant in slot 1 will not change the flash memory partition layout for an application, as the slot is already taken into account by default.
   Instead, the flash memory content of slot 1 will be populated by an image instead of being erased or left as-is.

.. _ug_bootloader_flash:

Flash memory partitions
=======================

Flash memory partitioning is handled differently for each bootloader.

.. _ug_bootloader_flash_b0:

|NSIB| partitions
-----------------

See :ref:`bootloader_flash_layout` for implementation-specific information about this bootloader.

.. _ug_bootloader_flash_mcuboot:

MCUboot partitions
------------------

> MCUboot requires two image slots:

Reviewer 5: @kyberdin

> MCUboot requires two image slots:

Author: @MarekPieta  Added a mention about single-slots for memory-constrained devices. Please see latest commit.
